numba/np/arraymath.py

```diff
@@ -4146,24 +4149,25 @@ def impl(ap1, ap2, mode, direction):
         else:
             raise ValueError("Invalid direction")
 
-        for i in range(n_left):
-            ret[idx] = innerprod(ap1[:idx + 1], ap2[-(idx + 1):])
+        for i in range(n - n_left, n):
+            ret[idx] = innerprod(ap1[:i], ap2[-i:])
```

Comment:
I would recommend keeping this particular part of implementation closer to NumPy if possible. It serves as a reference point for most of the users, even if it is a bit less readable/natural. 

As for the optimization of the increment being made over here, the constant increment within the loop (`+1` part) as a general rule of thumb is supposed to be optimized out by LLVM/compiler passes. (Although I'll need to check for this before making this a claim.)

Reply:
This is the best I can do, without making it a total mess (in my opinion, making it closer to Numpy's implementation would dramatically decrease readability, and maybe performances):

https://github.com/numba/numba/blob/0c5e47ee43053d52c2e526ecd7cfe702e9c9be79/numba/np/arraymath.py#L4146-L4158